File: chesscomp/analysis.py
"""Structured analysis of a problem: phases (set play, tries, solution), threats,
variations, refutations, and cross-phase relations (changed / transferred / reciprocal mates).

Everything returned is plain dicts/lists (JSON-serialisable) so an AI agent can reason over it.
"""
from __future__ import annotations
import time
import chess
from .core import Problem, san
from .solver import Engine, Budget


_ORIGIN = {}   # current square -> diagram square, for the piece that made the first move


def _mid(board, m):
    """Problemist identity of a move: piece + its diagram square + destination (+promotion).
    So Qb8-g8 after 1.Bh2 and Qh8-g8 after 1.Qh8 are the same mate 'Qg8' by the same queen."""
    p = board.piece_at(m.from_square)
    frm = chess.square_name(m.from_square)
    frm = _ORIGIN.get(frm, frm)
    promo = chess.piece_symbol(m.promotion) if m.promotion else ''
    return f"{'KQRBSP'[6 - p.piece_type] if p else '?'}{frm}{chess.square_name(m.to_square)}{promo}"


def _cont_list(board, moves):
    return [{'uci': m.uci(), 'id': _mid(board, m), 'san': san(board, m)} for m in moves]


def _variations(eng: Engine, board: chess.Board, n: int, mode: str, threat_ucis: set):
    """Board: Black to move. For each Black move -> White continuations within n-1."""
    out = []
    for b in list(board.legal_moves):
        d_san = san(board, b); d_id = _mid(board, b)
        board.push(b)
        try:
            gb = eng.goal_after_black(board, mode)
            if gb is True:           # (selfmate) Black's move itself is the mate
                conts, ends = [], True
            elif gb is False:
                conts, ends = [], False
            else:
                conts = eng.continuations(board, n - 1, mode) if n - 1 >= 1 else []
                ends = False
            cl = _cont_list(board, conts)
        finally:
            board.pop()
        cu = {c['id'] for c in cl}
        out.append({
            'defence': {'uci': b.uci(), 'id': d_id, 'san': d_san},
            'continuations': cl,
            'goal_by_defence': ends,
            'refutes': (not ends) and not cl,
            'dual': len(cl) > 1,
            'threat_repeat': bool(threat_ucis) and bool(cu & threat_ucis),
        })
    return out


def _phase(eng, board, n, mode, first_move, kind, refs=None):
    """board is the position BEFORE first_move (White to move)."""
    ph = {'type': kind, 'first_move': {'uci': first_move.uci(), 'id': _mid(board, first_move), 'san': san(board, first_move)}}
    _ORIGIN.clear(); _ORIGIN[chess.square_name(first_move.to_square)] = chess.square_name(first_move.from_square)
    board.push(first_move)
    try:
        if refs is not None:
            ph['refutations'] = _cont_list(board, refs)
        threat = []
        if not board.is_check() and n - 1 >= 1:
            board.push(chess.Move.null())
            try:
                threat = _cont_list(board, eng.continuations(board, n - 1, mode))
            finally:
                board.pop()
        ph['threat'] = threat
        ph['zugzwang'] = not threat and not board.is_check()
        ph['check_key'] = board.is_check()
        ph['variations'] = _variations(eng, board, n, mode, {t['id'] for t in threat})
        ph['flights'] = _flights(board)
    finally:
        board.pop()
        _ORIGIN.clear()
    return ph


def _flights(board: chess.Board):
    """Squares the black king can move to (Black to move)."""
    k = board.king(chess.BLACK)
    return sorted(chess.square_name(m.to_square) for m in board.legal_moves if m.from_square == k)


def _set_play(eng, board, n, mode):
    if n - 1 < 1 and mode == 'direct':
        return None
    b = board.copy(stack=False)
    b.turn = chess.BLACK
    b.ep_square = None
    if not b.is_valid() or b.is_check():
        return None
    ph = {'type': 'set', 'first_move': None, 'threat': [], 'zugzwang': False}
    ph['variations'] = _variations(eng, b, n, mode, set())
    ph['flights'] = _flights(b)
    ph['unprovided'] = [v['defence']['san'] for v in ph['variations'] if v['refutes']]
    return ph


def _defence_map(ph, unique_only=True):
    """black move uci -> set of continuation ucis, for real defences (threat defeated)."""
    return {v['defence']['id']: {c['id'] for c in v['continuations']}
            for v in ph['variations']
            if v['continuations'] and not v['threat_repeat'] and (len(v['continuations']) == 1 or not unique_only)}


def _label(ph):
    if ph['type'] == 'set':
        return 'set play'
    return f"{ph['type']} {ph['first_move']['san']}"


def compare_phases(phases, board=None):
    """Detect changed, transferred and reciprocally changed continuations between phases."""
    rel = {'changed': [], 'transferred': [], 'reciprocal': [], 'unchanged': []}
    sanmap = {}
    for ph in phases:
        for v in ph['variations']:
            sanmap.setdefault(v['defence']['id'], v['defence']['san'])
            for c in v['continuations']:
                sanmap.setdefault(c['id'], c['san'])
        for t in ph.get('threat', []):
            sanmap.setdefault(t['id'], t['san'])
    for i in range(len(phases)):
        for j in range(i + 1, len(phases)):
            A, B = phases[i], phases[j]
            ma, mb = _defence_map(A), _defence_map(B)
            common = set(ma) & set(mb)
            for d in sorted(common):
                if ma[d].isdisjoint(mb[d]):
                    rel['changed'].append({'defence': sanmap[d], 'phases': [_label(A), _label(B)],
                                           'from': sorted(sanmap[x] for x in ma[d]),
                                           'to': sorted(sanmap[x] for x in mb[d])})
                else:
                    rel['unchanged'].append({'defence': sanmap[d], 'phases': [_label(A), _label(B)]})
            # transferred: continuation after defence d in A appears after a different defence e in B
            for d, ca in ma.items():
                for e, cb in mb.items():
                    if d != e and d[1:3] != e[1:3]:
                        for w in ca & cb:
                            if w not in mb.get(d, set()):
                                rel['transferred'].append({'continuation': sanmap[w], 'phases': [_label(A), _label(B)],
                                                           'defences': [sanmap[d], sanmap[e]]})
            # reciprocal: d1->X, d2->Y  becomes  d1->Y, d2->X
            cl = sorted(common)
            for x in range(len(cl)):
                for y in range(x + 1, len(cl)):
                    d1, d2 = cl[x], cl[y]
                    if (ma[d1] & mb[d2]) and (ma[d2] & mb[d1]) and ma[d1].isdisjoint(mb[d1]) and ma[d2].isdisjoint(mb[d2]):
                        rel['reciprocal'].append({'defences': [sanmap[d1], sanmap[d2]], 'phases': [_label(A), _label(B)]})
    rel['patterns'] = detect_patterns(phases, sanmap, board)
    return rel


def _line_dir(a, b):
    fa, ra, fb, rb = chess.square_file(a), chess.square_rank(a), chess.square_file(b), chess.square_rank(b)
    df, dr = fb - fa, rb - ra
    if df == 0 and dr == 0:
        return None
    if df == 0 or dr == 0 or abs(df) == abs(dr):
        return ((df > 0) - (df < 0), (dr > 0) - (dr < 0))
    return None


def _between(a, b, c):
    """Is square b strictly between a and c on one line?"""
    d1, d2 = _line_dir(a, b), _line_dir(b, c)
    return d1 is not None and d1 == d2


def bristol(first_uci, follow_uci):
    """Bristol clearance (taught by E. Bourd, 2026): the first move travels along a line A->B; a second
    piece then moves along the SAME line in the SAME direction, passing over A and stopping short of B."""
    A, B = chess.parse_square(first_uci[:2]), chess.parse_square(first_uci[2:4])
    C, D = chess.parse_square(follow_uci[:2]), chess.parse_square(follow_uci[2:4])
    d = _line_dir(A, B)
    return d is not None and _line_dir(C, D) == d and _between(C, A, D) and (D == B or _between(A, D, B)) and D != B


def cut_lines(board_before: chess.Board, move: chess.Move):
    """White line pieces whose lines are CLOSED by `move` (the moving piece stands in their way afterwards).
    Returns {square_cut: [white piece names whose guard on it is lost]}."""
    after = board_before.copy(stack=False); after.push(move)
    ghost = after.copy(stack=False); ghost.remove_piece_at(move.to_square)
    cut = {}
    for sq in chess.SquareSet(after.occupied_co[chess.WHITE]):
        if sq == move.to_square or after.piece_type_at(sq) not in (chess.QUEEN, chess.ROOK, chess.BISHOP):
            continue
        lost = int(ghost.attacks(sq)) & ~int(after.attacks(sq))
        for t in chess.SquareSet(lost):
            if t != move.to_square:
                p = after.piece_at(sq)
                cut.setdefault(chess.square_name(t), []).append(f"{'QRB'[[chess.QUEEN, chess.ROOK, chess.BISHOP].index(p.piece_type)]}{chess.square_name(sq)}")
    return cut


def interference_motive(board: chess.Board, try_ph: dict, key_ph: dict | None = None):
    """Why does the try fail? If the first move closed a White line and the refutation exploits a square
    that line used to guard, return the explanation (white self-interference try)."""
    fm = chess.Move.from_uci(try_ph['first_move']['uci'])
    cut = cut_lines(board, fm)
    if not cut:
        return None
    out = []
    for r in try_ph.get('refutations', []):
        b = board.copy(stack=False); b.push(fm); rm = chess.Move.from_uci(r['uci'])
        hits = set()
        if chess.square_name(rm.to_square) in cut:
            hits.add(chess.square_name(rm.to_square))
        b.push(rm)
        # candidate White answers: the try's threat, and the answer this defence gets in the SOLUTION
        cands = list(try_ph.get('threat', []))
        if key_ph:
            for v in key_ph['variations']:
                if v['defence']['uci'] == r['uci']:
                    cands += v['continuations']
        for t in cands:
            tm = chess.Move.from_uci(t['uci'])
            if tm not in b.legal_moves:
                continue
            b2 = b.copy(stack=False); b2.push(tm)
            for m in b2.legal_moves:          # how does Black survive the threat move?
                if chess.square_name(m.to_square) in cut:
                    hits.add(chess.square_name(m.to_square))
        if hits:
            out.append({'refutation': r['san'], 'squares': sorted(hits),
                        'lines': sorted({x for h in hits for x in cut[h]})})
    return out or None


def detect_patterns(phases, sanmap, board=None):
    """Named patterns. Extend this as new themes are taught."""
    pats = []
    if board is not None:
        for ph in phases:
            if ph['type'] == 'try':
                keyp = next((q for q in phases if q['type'] == 'key'), None)
                im = interference_motive(board, ph, keyp)
                if im:
                    for x in im:
                        pats.append({'name': 'White interference try', 'phases': [_label(ph)],
                                     'defence': f"1.{ph['first_move']['san']}? cuts {'/'.join(x['lines'])} -> {','.join(x['squares'])}; 1...{x['refutation']}!"})
    for ph in phases:
        if ph['type'] not in ('try', 'key'):
            continue
        fm = ph['first_move']['uci']
        for t in ph['threat']:
            if bristol(fm, t['uci']):
                pats.append({'name': 'Bristol', 'phases': [_label(ph)], 'defence': f"1.{ph['first_move']['san']} + 2.{t['san']}"})
        for v in ph['variations']:
            for c in v['continuations']:
                if not v['threat_repeat'] and bristol(fm, c['uci']):
                    pats.append({'name': 'Bristol (in variation)', 'phases': [_label(ph)],
                                 'defence': f"1.{ph['first_move']['san']} {v['defence']['san']} 2.{c['san']}"})
    real = [p for p in phases if p['type'] in ('try', 'key')]
    for i in range(len(real)):
        for j in range(len(real)):
            if i == j:
                continue
            A, B = real[i], real[j]
            ta = {t['id'] for t in A['threat']}; tb = {t['id'] for t in B['threat']}
            if not ta or not tb:
                continue
            if i < j and ta.isdisjoint(tb):
                pats.append({'name': 'changed threat', 'phases': [_label(A), _label(B)],
                             'threats': [sorted(sanmap.get(x, x) for x in ta), sorted(sanmap.get(x, x) for x in tb)]})
            ma, mb = _defence_map(A), _defence_map(B)
            # Le Grand: threat A / variation d->B  becomes  threat B / d->A
            for d in set(ma) & set(mb):
                if ma[d] <= tb and mb[d] <= ta and ta.isdisjoint(tb) and i < j:
                    pats.append({'name': 'Le Grand', 'phases': [_label(A), _label(B)], 'defence': sanmap.get(d, d)})
            # pseudo Le Grand: threat of A appears as mate after a defence in B, and vice versa (different defences)
            if i < j:
                xa = [d for d, c in mb.items() if c <= ta]
                xb = [d for d, c in ma.items() if c <= tb]
                if xa and xb and ta.isdisjoint(tb) and not (set(xa) & set(xb)):
                    pats.append({'name': 'Pseudo Le Grand', 'phases': [_label(A), _label(B)],
                                 'defences': [sanmap.get(xb[0]), sanmap.get(xa[0])]})
    # Threat reversal: 1.A? threat B  /  1.B! threat A  (first move and threat swap)
    for i in range(len(real)):
        for j in range(i + 1, len(real)):
            A, B = real[i], real[j]
            ta = {t['id'] for t in A['threat']}; tb = {t['id'] for t in B['threat']}
            if A['first_move'].get('id') in tb and B['first_move'].get('id') in ta:
                pats.append({'name': 'Threat reversal', 'phases': [_label(A), _label(B)]})
    # Dombrovskis: in phase A threat X is defeated by defence d (d refutes, or d stops X);
    # in phase B the same defence d is answered by X.
    for A in real:
        ta = {t['id'] for t in A['threat']}
        if not ta:
            continue
        stoppers = {v['defence']['id'] for v in A['variations'] if not v['threat_repeat']}
        for B in real:
            if B is A:
                continue
            for d, conts in _defence_map(B).items():
                if d in stoppers and conts & ta:
                    refuting = d in {r.get('id') for r in A.get('refutations', [])}
                    pats.append({'name': 'Dombrovskis' + (' (refutation)' if refuting else ''),
                                 'phases': [_label(A), _label(B)], 'defence': sanmap.get(d, d)})
    # Zagoruiko: best subset of >=3 phases sharing >=2 defences whose continuations change in every phase
    from itertools import combinations
    maps = [(p, _defence_map(p)) for p in phases]
    best = None
    for k in range(min(len(maps), 6), 2, -1):
        for combo in combinations(maps, k):
            common = set.intersection(*[set(m) for _, m in combo])
            steady = [d for d in common
                      if all(combo[x][1][d].isdisjoint(combo[y][1][d]) for x in range(k) for y in range(x + 1, k))]
            if len(steady) >= 2 and (best is None or (k, len(steady)) > (best[0], len(best[1]))):
                best = (k, steady, [_label(p) for p, _ in combo])
        if best:
            break
    if best:
        pats.append({'name': f'Zagoruiko {best[0]}x{len(best[1])}', 'defences': sorted(sanmap.get(d, d) for d in best[1]),
                     'phases': best[2]})
    return pats


def analyse(problem: Problem, max_refutations: int = 1, include_tries: bool = True,
            include_set: bool = True, time_limit: float | None = 120, checking_tries: bool = False) -> dict:
    st = problem.stip
    board = problem.board.copy()
    eng = Engine(time_limit)
    t0 = time.time()
    res = {'stipulation': st.text, 'kind': st.kind, 'fen': board.fen(), 'count': problem.count()}
    try:
        if st.kind == 'help':
            plies = 2 * st.full_moves + (1 if st.half else 0)
            sols = eng.help_solutions(board, plies)
            res['solutions'] = [_help_line(board, s, st) for s in sols]
            res['n_solutions'] = len(sols)
        else:
            n, mode = st.full_moves, st.kind
            phases = []
            if include_set:
                sp = _set_play(eng, board, n, mode)
                if sp and any(v['continuations'] for v in sp['variations']):
                    phases.append(sp)
            keys, tries = [], []
            for w in eng.ordered(board):
                board.push(w)
                try:
                    refs = eng.refutations(board, n, mode, limit=max_refutations if include_tries else 0)
                finally:
                    board.pop()
                if refs is None:
                    continue
                if not refs:
                    keys.append(w)
                elif include_tries and len(refs) <= max_refutations and (checking_tries or not board.gives_check(w)):
                    tries.append((w, refs))
            for w, refs in tries:
                phases.append(_phase(eng, board, n, mode, w, 'try', refs))
            for w in keys:
                phases.append(_phase(eng, board, n, mode, w, 'key', []))
            res['keys'] = [san(board, k) for k in keys]
            res['cooked'] = len(keys) > 1
            res['phases'] = phases
            res['relations'] = compare_phases(phases, board)
            try:
                from .motives import explain_key_phase, detect_grimshaw
                ex = explain_key_phase(problem, res)
                res['motives'] = ex
                for g in detect_grimshaw(ex):
                    res['relations']['patterns'].append({'name': 'Grimshaw', 'phases': ['key'], 'defence': f"{'/'.join(g['defences'])} on {g['square']}"})
            except Exception as exc:
                res['motives_error'] = repr(exc)
    except Budget:
        res['timeout'] = True
    res['nodes'] = eng.nodes
    res['seconds'] = round(time.time() - t0, 2)
    return res


def _help_line(board, moves, st):
    """Helpmate line in album numbering: h#n -> 1.B W 2.B W# ; h#n.5 -> 1...W 2.B W ..."""
    b = board.copy(stack=False)
    out, n = [], 1
    for i, m in enumerate(moves):
        s = san(b, m)
        if st.half and i == 0:
            out.append(f'1...{s}'); n = 2
        elif b.turn == chess.BLACK:
            out.append(f'{n}.{s}')
        else:
            out.append(s); n += 1
        b.push(m)
    return {'line': ' '.join(out), 'uci': [m.uci() for m in moves]}

File: docs/analysis.js
/* chesscomp-web: #2 / s#2 / h#2 analysis in the browser.
   Mirrors chesscomp/analysis.py: set play, tries with refutations, key, variations, duals,
   changed / reciprocal mates, and the dual-avoidance motive.
   Depends on chess.js (Chess class). */
(function (root, factory) {
  if (typeof module === 'object' && module.exports) module.exports = factory(require('chess.js').Chess);
  else root.ChessComp = factory(root.Chess);
}(typeof self !== 'undefined' ? self : this, function (Chess) {

  const S = s => s.replace(/^N/, 'S').replace('=N', '=S').replace('O-O-O', '0-0-0').replace('O-O', '0-0');

  function boardFrom(fen, turn) {
    const parts = fen.trim().split(/\s+/);
    const placement = parts[0];
    const f = `${placement} ${turn} - - 0 1`;
    try { return new Chess(f); } catch (e) { return null; }
  }

  function legal(c) { return c.moves({ verbose: true }); }

  function matesIn1(c) {                       // side to move mates at once
    const out = [];
    for (const m of legal(c)) {
      c.move(m);
      if (c.isCheckmate()) out.push(m);
      c.undo();
    }
    return out;
  }

  // identity of a move: piece + origin + destination, so the same unit is tracked across phases
  const mid = m => `${(m.piece || '').toUpperCase()}${m.from}${m.to}${m.promotion || ''}`;

  /* Does White force mate within n moves? (direct mates only, n = 1 or 2) */
  function whiteMatesIn2(c) {
    for (const w of legal(c)) {
      c.move(w);
      const ok = blackAllAnswered(c);
      c.undo();
      if (ok) return true;
    }
    return false;
  }

  function blackAllAnswered(c) {               // Black to move: is every reply mated next move?
    if (c.isCheckmate()) return true;
    if (c.isStalemate() || c.isInsufficientMaterial()) return false;
    for (const b of legal(c)) {
      c.move(b);
      const dead = c.isGameOver() && !c.isCheckmate();
      const ms = dead ? [] : matesIn1(c);
      c.undo();
      if (!ms.length) return false;
    }
    return true;
  }

  function solvesShort(c, mv) {                // a first move that mates at once cooks a #2
    c.move(mv);
    const short = c.isCheckmate();
    c.undo();
    return short;
  }

  /* Analyse one first move: threat, variations, duals. */
  function phaseAfter(c, first, kind) {
    const sanFirst = S(c.move(first).san);
    const ph = { type: kind, first: sanFirst, firstId: mid(first), from: first.from, to: first.to,
                 threat: [], variations: [], refutations: [] };
    // threat: what White plays if Black does nothing
    const nullFen = c.fen().replace(/ b /, ' w ');
    let nc = null;
    try { nc = new Chess(nullFen); } catch (e) { nc = null; }
    if (nc && !nc.isCheck()) ph.threat = matesIn1(nc).map(m => ({ san: S(m.san), id: mid(m) }));
    const threatIds = new Set(ph.threat.map(t => t.id));
    for (const d of legal(c)) {
      c.move(d);
      const dead = c.isGameOver() && !c.isCheckmate();
      const ms = dead ? [] : matesIn1(c);
      const conts = ms.map(m => ({ san: S(m.san), id: mid(m), from: m.from, to: m.to, uci: m.from + m.to }));
      c.undo();
      const ids = new Set(conts.map(x => x.id));
      const repeats = ph.threat.length > 0 && [...ids].some(x => threatIds.has(x));
      const v = { defence: S(d.san), defFrom: d.from, defTo: d.to, uci: d.from + d.to,
                  conts, dual: conts.length > 1, refutes: conts.length === 0, threatRepeat: repeats };
      ph.variations.push(v);
      if (v.refutes) ph.refutations.push(S(d.san));
    }
    c.undo();
    return ph;
  }

  function setPlay(fen) {
    const c = boardFrom(fen, 'b');
    if (!c || c.isCheck() || c.isGameOver()) return null;
    const ph = { type: 'set', first: null, threat: [], variations: [], unprovided: [] };
    for (const d of legal(c)) {
      c.move(d);
      const dead = c.isGameOver() && !c.isCheckmate();
      const ms = dead ? [] : matesIn1(c);
      const conts = ms.map(m => ({ san: S(m.san), id: mid(m) }));
      c.undo();
      if (!conts.length) ph.unprovided.push(S(d.san));
      else ph.variations.push({ defence: S(d.san), uci: d.from + d.to, conts, dual: conts.length > 1,
                                threatRepeat: false, refutes: false });
    }
    return ph;
  }

  /* changed / reciprocal continuations between phases (real defences only, unique mates) */
  function defMap(ph) {
    const m = {};
    for (const v of ph.variations)
      if (v.conts.length === 1 && !v.threatRepeat) m[v.uci] = { id: v.conts[0].id, san: v.conts[0].san, def: v.defence };
    return m;
  }
  function label(ph) { return ph.type === 'set' ? 'set play' : `${ph.type} ${ph.first}`; }

  function relations(phases) {
    const changed = [], reciprocal = [];
    for (let i = 0; i < phases.length; i++)
      for (let j = i + 1; j < phases.length; j++) {
        const A = defMap(phases[i]), B = defMap(phases[j]);
        const common = Object.keys(A).filter(k => k in B);
        for (const d of common)
          if (A[d].id !== B[d].id)
            changed.push({ defence: A[d].def, from: A[d].san, to: B[d].san, phases: [label(phases[i]), label(phases[j])] });
        for (let x = 0; x < common.length; x++)
          for (let y = x + 1; y < common.length; y++) {
            const d1 = common[x], d2 = common[y];
            if (A[d1].id === B[d2].id && A[d2].id === B[d1].id && A[d1].id !== B[d1].id)
              reciprocal.push({ defences: [A[d1].def, A[d2].def], phases: [label(phases[i]), label(phases[j])] });
          }
      }
    return { changed, reciprocal };
  }

  /* Why does each sibling mate fail after a given defence? (dual-avoidance motive) */
  function dualAvoidance(fen, keyMove) {
    const c = boardFrom(fen, 'w');
    if (!c) return [];
    c.move(keyMove);
    const nullFen = c.fen().replace(/ b /, ' w ');
    let threatIds = new Set();
    try { const nc = new Chess(nullFen); threatIds = new Set(matesIn1(nc).map(mid)); } catch (e) {}
    const defs = [];
    for (const d of legal(c)) {
      c.move(d);
      const ms = matesIn1(c);
      c.undo();
      if (ms.length === 1 && !threatIds.has(mid(ms[0]))) defs.push({ d, m: ms[0] });
    }
    const out = [];
    for (const { d, m } of defs) {
      const item = { defence: S(d.san), mate: S(m.san), avoided: [] };
      for (const other of defs) {
        if (other.m.from === m.from && other.m.to === m.to) continue;
        c.move(d);
        const cand = legal(c).find(x => x.from === other.m.from && x.to === other.m.to && (x.promotion || '') === (other.m.promotion || ''));
        if (!cand) { c.undo(); continue; }
        const osan = S(cand.san);
        c.move(cand);
        if (c.isCheckmate()) { item.avoided.push({ mate: osan, refutation: null, kind: 'DUAL' }); c.undo(); c.undo(); continue; }
        const escapes = legal(c);
        let best = null;
        for (const e of escapes) {
          let kind;
          if (e.to === cand.to) kind = e.from === d.to ? 'defender retains control' : 'another unit captures the mating unit';
          else if (e.piece === 'k') kind = 'king flight';
          else if (!c.isCheck()) kind = 'no check';
          else kind = 'interposition';
          const rank = ['defender retains control', 'another unit captures the mating unit', 'interposition', 'king flight', 'no check'].indexOf(kind);
          if (!best || rank < best.rank) best = { rank, kind, san: S(e.san) };
        }
        if (best) item.avoided.push({ mate: osan, refutation: best.san, kind: best.kind });
        c.undo(); c.undo();
      }
      out.push(item);
    }
    return out;
  }

  /* main entry */
  function analyse(fen, stip, opts) {
    opts = opts || {};
    stip = (stip || '#2').toLowerCase().replace(/\s/g, '');
    const res = { fen: fen.trim().split(/\s+/)[0], stip, phases: [], keys: [], cooked: false, relations: { changed: [], reciprocal: [] } };
    const t0 = Date.now();

    if (stip === 'h#2' || stip === 'h#1.5' || stip === 'h#2.5') {
      const plies = stip === 'h#2' ? 4 : (stip === 'h#1.5' ? 3 : 5);
      res.solutions = helpSolutions(boardFrom(fen, stip.endsWith('.5') ? 'w' : 'b'), plies);
      res.ms = Date.now() - t0;
      return res;
    }

    const c = boardFrom(fen, 'w');
    if (!c) { res.error = 'bad FEN'; return res; }
    if (c.isCheck()) { res.error = 'Black king is in check in the diagram'; return res; }

    const sp = setPlay(fen);
    if (sp && sp.variations.some(v => v.conts.length)) res.phases.push(sp);

    const keys = [], tries = [];
    for (const w of legal(c)) {
      if (solvesShort(c, w)) { res.keys.push(S(c.move(w).san)); c.undo(); res.cooked = true; continue; }
      c.move(w);
      const ok = blackAllAnswered(c);
      let refCount = 0;
      if (!ok) {
        let threat = false;
        const nullFen = c.fen().replace(/ b /, ' w ');
        try { const nc = new Chess(nullFen); threat = !nc.isCheck() && matesIn1(nc).length > 0; } catch (e) {}
        if (threat) {
          for (const b of legal(c)) {
            c.move(b);
            const dead = c.isGameOver() && !c.isCheckmate();
            const ms = dead ? [] : matesIn1(c);
            c.undo();
            if (!ms.length) refCount++;
            if (refCount > (opts.maxRefutations || 1)) break;
          }
        } else refCount = 99;
      }
      c.undo();
      if (ok) keys.push(w);
      else if (refCount <= (opts.maxRefutations || 1)) tries.push(w);
    }
    for (const t of tries) res.phases.push(phaseAfter(c, t, 'try'));
    for (const k of keys) res.phases.push(phaseAfter(c, k, 'key'));
    res.keys = res.keys.concat(keys.map(k => { const s = S(c.move(k).san); c.undo(); return s; }));
    res.cooked = res.cooked || res.keys.length > 1;
    res.relations = relations(res.phases);
    if (keys.length === 1) res.dualAvoidance = dualAvoidance(fen, keys[0]);
    res.ms = Date.now() - t0;
    return res;
  }

  function helpSolutions(c, plies, max) {
    max = max || 30;
    const sols = [], line = [];
    (function rec(p) {
      if (sols.length >= max) return;
      for (const m of legal(c)) {
        c.move(m); line.push(S(m.san));
        if (p === 1) { if (c.isCheckmate()) sols.push(line.slice()); }
        else if (!c.isGameOver()) rec(p - 1);
        c.undo(); line.pop();
      }
    })(plies);
    return sols.map(l => l.map((s, i) => (i % 2 === 0 ? `${i / 2 + 1}.${s}` : s)).join(' '));
  }

  function formatReport(res) {
    if (res.error) return res.error;
    const L = [];
    if (res.solutions) {
      L.push(`${res.solutions.length} solution(s):`);
      res.solutions.forEach(s => L.push('   ' + s));
      return L.join('\n');
    }
    for (const ph of res.phases) {
      if (ph.type === 'set') {
        L.push('Set play:');
      } else {
        let h = `${ph.type === 'key' ? 'Key' : 'Try'}: 1.${ph.first}${ph.type === 'key' ? '!' : '?'}`;
        if (ph.threat.length) h += ` (threat: 2.${ph.threat.map(t => t.san).join('/')})`;
        else h += ' (zugzwang or check)';
        L.push(h);
      }
      const groups = new Map();
      let hidden = 0;
      for (const v of ph.variations) {
        if (v.refutes) continue;
        if (v.threatRepeat) { hidden++; continue; }
        const k = v.uci.slice(0, 2) + '|' + v.conts.map(x => x.san).join('/');
        if (!groups.has(k)) groups.set(k, { defs: [], conts: v.conts, dual: v.dual });
        groups.get(k).defs.push(v.defence);
      }
      for (const g of groups.values())
        L.push(`   1...${g.defs.join('/')} 2.${g.conts.map(x => x.san).join('/')}${g.dual ? '  [DUAL]' : ''}`);
      if (hidden) L.push(`   (${hidden} further moves allow the threat)`);
      if (ph.type === 'try' && ph.refutations.length) L.push('   but ' + ph.refutations.map(r => `1...${r}!`).join(', '));
      if (ph.type === 'set' && ph.unprovided.length) L.push('   unprovided: ' + ph.unprovided.slice(0, 12).join(', '));
    }
    if (!res.keys.length) L.push('No solution found.');
    else if (res.cooked) L.push(`COOKED: ${res.keys.length} keys: ${res.keys.join(', ')}`);
    if (res.relations.changed.length) {
      L.push('Changed continuations:');
      res.relations.changed.forEach(c => L.push(`   1...${c.defence}: ${c.from} -> ${c.to}   [${c.phases[0]} vs ${c.phases[1]}]`));
    }
    if (res.relations.reciprocal.length) {
      L.push('Reciprocal changes:');
      res.relations.reciprocal.forEach(r => L.push(`   ${r.defences.join(' / ')}   [${r.phases[0]} vs ${r.phases[1]}]`));
    }
    if (res.dualAvoidance && res.dualAvoidance.length) {
      const marked = res.dualAvoidance.filter(x => x.avoided.length);
      if (marked.length) {
        L.push('Dual avoidance:');
        marked.forEach(it => L.push(`   1...${it.defence} 2.${it.mate}  ` +
          it.avoided.map(a => a.refutation ? `(2.${a.mate}? ${a.refutation}!  ${a.kind})` : `(2.${a.mate}? also mate - DUAL)`).join(' ')));
        const kinds = new Set(marked.map(it => it.avoided[0] && it.avoided[0].kind).filter(Boolean));
        if (kinds.size === 1) L.push(`   [unified avoidance: ${[...kinds][0]}]`);
      }
    }
    L.push(`(${res.ms} ms)`);
    return L.join('\n');
  }

  return { analyse, formatReport, setPlay, dualAvoidance };
}));
